### PR TITLE
arch/sim: Implement mm_mallinfo and malloc_size for the custom heap

### PR DESCRIPTION
**Pull request by xiaoxiang781216**
## Summary

- mm: Support malloc_size function
- arch/sim: Remove host_malloc and host_calloc
- arch/sim: Implement malloc_size for the custom heap
- arch/sim: Implement mm_mallinfo for the custom heap
- boards/sim: Add asan config to test the custom heap

## Impact
free can show correctly with the custom heap

## Testing
before change:
```
nsh> free
                     total       used       free    largest  nused  nfree
        Umem:            0          0          0          0      0      0
```
after change:
```
nsh> free
                     total       used       free    largest  nused  nfree
        Umem:            0      72288          0          0     32      0
```

### CODE REVIEW

File: drivers/note/notelog_driver.c [hunk 1]
```diff
@@ -189,3 +260,131 @@ void sched_note_csection(FAR struct tcb_s *tcb, bool enter)
...
+         tcb, spinlock);
+#endif
+#endif
+}
+#endif
+
+#ifdef CONFIG_SCHED_INSTRUMENTATION_SYSCALL
+void sched_note_syscall_enter(int nr, int argc, ...)
+{
+  char buf[256];
```
Comment (acassis) [inline drivers/note/notelog_driver.c:357]: @xiaoxiang781216 shouldn't this buffer size be configurable? Seams too big for low-end microcontrollers

**Comment by acassis** [comment]
@xiaoxiang781216 I think you need to rebase it, the  michallenc commit are mixed in your PR

File: drivers/note/notelog_driver.c (hunk 1 above)
Comment (xiaoxiang781216) [reply drivers/note/notelog_driver.c:357]: Ok, the buffer is changed to 128B, which is enough to hold 8 ponter dump.

**Comment by xiaoxiang781216** [comment]
> @xiaoxiang781216 I think you need to rebase it, the michallenc commit are mixed in your PR

Done, it's my fault to mix https://github.com/apache/incubator-nuttx/pull/3979 with this PR.

File: drivers/note/notelog_driver.c (hunk 1 above)
Comment (xiaoxiang781216) [reply drivers/note/notelog_driver.c]: The change apply in this PR: https://github.com/apache/incubator-nuttx/pull/3979

**Comment by xiaoxiang781216** [comment]
@acassis could you merge this PR?

File: include/malloc.h [hunk 2]
```diff
@@ -31,6 +31,8 @@
  * Pre-processor Definitions
  ****************************************************************************/
 
+#define malloc_usable_size malloc_size
+
```
Comment (davids5) [inline include/malloc.h:35]: @xiaoxiang781216 Why is this needed?

Comment (acassis) [reply include/malloc.h:35]: > @acassis could you merge this PR?

I'll wait your respond David about the malloc_size and then will merge it.

Comment (xiaoxiang781216) [reply include/malloc.h:35]: Since the standard doesn't define API to get the malloc size, It isn't unfortunate that popular Unix variant OS define two different function for it:

1. Linux and FreeBSD define malloc_usable_size:
    https://man7.org/linux/man-pages/man3/malloc_usable_size.3.html
    https://www.freebsd.org/cgi/man.cgi?query=malloc_usable_size&apropos=0&sektion=0&manpath=FreeBSD+7.1-RELEASE&format=html
2. macOS define malloc_size:
    https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man3/malloc_size.3.html

**Review by yamt — COMMENTED** [review]
i'm not sure if malloc_usable_size macro is a good idea. otherwise lgtm.

File: include/malloc.h (hunk 2 above)
Comment (xiaoxiang781216) [reply include/malloc.h:35]: But a simple macro can improve the portability without bad side effect.

Comment (davids5) [reply include/malloc.h:35]: @xiaoxiang781216 - How about adding a comment "For Linux and MacOS compatibility"

Comment (xiaoxiang781216) [reply include/malloc.h:35]: Done.

**Review by davids5 — APPROVED** [review]
@xiaoxiang781216 - Thank you. 
@yamt any other suggestions?

**Review by acassis — APPROVED** [review]

**Comment by yamt** [comment]
> @xiaoxiang781216 - Thank you.
> @yamt any other suggestions?

i have nothing to add to what i have already commented.
